artman/cli/main2.py

```diff
@@ -0,0 +1,435 @@
+        '--dryrun',
+        dest='dryrun',
+        action='store_true',
+        help='[Optional] When specified, artman will skip the remote '
+        'publishing step.', )
+    parser_publish.set_defaults(dryrun=False)
+
+    return parser.parse_args(args=args)
+
+
+def read_user_config(flags):
+    """Read the user config from disk and return it.
+
+    Args:
+        flags (argparse.Namespace): The flags from sys.argv.
+
+    Returns:
+        dict: The user config.
+    """
+    # Load the user configuration if it exists and save a dictionary.
+    user_config = {}
+    user_config_file = os.path.realpath(os.path.expanduser(flags.user_config))
+    if os.path.isfile(user_config_file):
+        with io.open(user_config_file) as ucf:
+            user_config = yaml.load(ucf.read(), Loader=yaml.Loader) or {}
+
+    # Sanity check: Is there a configuration? If not, abort.
+    if not user_config:
+        setup_logging(INFO)
+        logger.critical('No user configuration found.')
+        logger.warn('This is probably your first time running Artman.')
+        logger.warn('Run `configure-artman` to get yourself set up.')
+        sys.exit(64)
+
+    # Done; return the user config.
+    return user_config
+
+
+def normalize_flags(flags, user_config):
+    """Combine the argparse flags and user configuration together.
+
+    Args:
+        flags (argparse.Namespace): The flags parsed from sys.argv
+        user_config (dict): The user configuration taken from
+                            ~/.artman/config.yaml.
+
+    Returns:
+        tuple (str, dict): 2-tuple containing:
+            - pipeline name
+            - pipeline arguments
+    """
+    flags.input_dir = os.path.abspath(flags.input_dir)
+    flags.output_dir = os.path.abspath(flags.output_dir)
+    flags.config = os.path.abspath(flags.config)
+    pipeline_args = {}
+
+    # Determine logging verbosity and then set up logging.
+    verbosity = support.resolve('verbosity', user_config, flags, default=INFO)
+    setup_logging(verbosity)
+
+    # Save local paths, if applicable.
+    # This allows the user to override the path to api-client-staging or
+    # toolkit on his or her machine.
+    pipeline_args['local_paths'] = support.parse_local_paths(
+        user_config, flags.input_dir)
+
+    artman_config_path = flags.config
+    if not os.path.isfile(artman_config_path):
+        logger.error(
+            'Artman config file `%s` doesn\'t exist.' % artman_config_path)
+        sys.exit(96)
+
+    try:
+        artifact_config = loader.load_artifact_config(
+            artman_config_path, flags.artifact_name, flags.input_dir)
+    except ValueError as ve:
+        logger.error('Artifact config loading failed with `%s`' % ve)
+
+    # If we were given just an API or BATCH, then expand it into the --config
+    # syntax.
+    shared_config_name = 'common.yaml'
+    if artifact_config.language == config_pb2.Artifact.RUBY:
+        # TODO(ethanbao): Figure out why.
```

Comment:
I am guessing this is probably because Ruby has an additional step to generate the bogus alternate pb2s so that yard can generate documentation for them.

In a perfect world, I would like to use protoc-docs-plugin to solve this, but the Ruby protobuf output lacks insertion points.